temporal101/Exercise2.hs

```diff
@@ -0,0 +1,147 @@
+module Main where
+
+import Control.Monad.IO.Class (liftIO)
+import Control.Monad.Logger (defaultOutput, runStdoutLoggingT)
+import Control.Monad.Trans.Reader (runReaderT)
+import Data.Aeson (FromJSON, ToJSON)
+import Data.Text (Text)
+import Data.UUID qualified as UUID
+import Data.UUID.V4 qualified as UUID.V4
+import DiscoverInstances (discoverInstances)
+import GHC.Generics (Generic)
+import RequireCallStack (RequireCallStack, provideCallStack)
+import System.IO (stdout)
+import Temporal.Activity (Activity)
+import Temporal.Client (mkWorkflowClientConfig, workflowClient)
+import Temporal.Client qualified as Client
+import Temporal.Core.Client (connectClient, defaultClientConfig)
+import Temporal.Duration (seconds)
+import Temporal.Payload (JSON (JSON))
+import Temporal.Runtime (TelemetryOptions (..), initializeRuntime)
+import Temporal.TH (WorkflowFn, ActivityFn)
+import Temporal.TH qualified
+import Temporal.Worker (Worker, WorkerConfig, startWorker)
+import Temporal.Worker qualified as Worker
+import Temporal.Workflow (Workflow, WorkflowId (..))
+import Temporal.Workflow qualified as Workflow
+import UnliftIO.Exception (bracket)
+
+-- | Arguments that shall be passed to 'composeGreeting'.
+--
+-- While we could write an 'Activity' that accepts multiple arguments in the
+-- in the same manner as any normal Haskell function, it's strongly encouraged
+-- to use standalone types that can have fields added to them in a
+-- backwards-compatible fashion.
+data ComposeGreetingInput = ComposeGreetingInput
+  { greeting :: Text,
+    name :: Text
+  }
+  deriving stock (Generic, Show)
+  deriving anyclass (FromJSON, ToJSON)
```

Comment:
may be worth a callout here that deriving ser/de instances via `Generic` means that changing record structure could break backwards compatibility if you're not careful.